File: rust/src/tracking/affordance.rs
use ndarray::Array1;

/// Spectral affordance field for real-time a cappella listening.
/// Ported and adapted from the Python/MATLAB research implementation.
pub struct AffordanceField {
    pub freqs: Array1<f64>,
    
    // State for persistence and change
    prev_presence: Array1<f64>,
    persistence: Array1<f64>,
    smooth_presence: Array1<f64>,
    
    // Hyperparameters
    persistence_alpha: f64,
    change_alpha: f64,
    masking_floor_db: f64,
    masking_spread_erb: f64,
    dominance_low_hz: f64,
    dominance_high_hz: f64,
    dominance_weight: f64,
    continuity_erb_sigma: f64,
}

impl AffordanceField {
    pub fn new(sample_rate: f64, frame_size: usize) -> Self {
        let n_bins = frame_size / 2 + 1;
        let freqs = Array1::from_shape_fn(n_bins, |i| {
            i as f64 * sample_rate / frame_size as f64
        });
        
        // Time constants (assuming ~23ms hop at 44.1kHz / 1024)
        let persistence_alpha = 0.95; 
        let change_alpha = 0.8;

        Self {
            freqs,
            prev_presence: Array1::zeros(n_bins),
            persistence: Array1::zeros(n_bins),
            smooth_presence: Array1::zeros(n_bins),
            persistence_alpha,
            change_alpha,
            masking_floor_db: -60.0,
            masking_spread_erb: 1.5,
            dominance_low_hz: 500.0,
            dominance_high_hz: 2000.0,
            dominance_weight: 1.0,
            continuity_erb_sigma: 0.5,
        }
    }

    /// Update the affordance field with a new magnitude spectrum frame.
    pub fn update(&mut self, mag: &Array1<f64>) -> Array1<f64> {
        let n = mag.len();
        
        // 1. Feature Presence (normalized magnitude)
        let max_mag = mag.fold(0.0_f64, |a, &b| a.max(b));
        let presence = if max_mag > 0.0 { 
            mag.mapv(|x| x / max_mag) 
        } else { 
            Array1::zeros(n) 
        };

        // 2. Peripheral Availability (Masking + Dominance)
        let mut availability = Array1::zeros(n);
        let mag_db = mag.mapv(|x| 20.0 * (x + 1e-12).log10());
        let peak_db = mag_db.fold(f64::NEG_INFINITY, |a, b| a.max(*b));
        
        for i in 0..n {
            let rel = mag_db[i] - peak_db;
            let val = (rel - self.masking_floor_db) / (0.0 - self.masking_floor_db);
            availability[i] = val.clamp(0.0, 1.0);
        }

        // Apply ERB smoothing to availability
        availability = self.smooth_along_erb(&availability, self.masking_spread_erb);

        // Dominance weight
        for i in 0..n {
            if self.freqs[i] >= self.dominance_low_hz && self.freqs[i] <= self.dominance_high_hz {
                availability[i] *= 1.0 + self.dominance_weight;
            }
        }
        
        // Normalize availability
        let max_avail = availability.fold(0.0_f64, |a, &b| a.max(b));
        if max_avail > 0.0 { 
            availability.mapv_inplace(|x| x / max_avail); 
        }

        // 3. Persistence (Exponential Moving Average)
        for i in 0..n {
            self.persistence[i] = self.persistence_alpha * self.persistence[i] + (1.0 - self.persistence_alpha) * presence[i];
        }
        let max_pers = self.persistence.fold(0.0_f64, |a, &b| a.max(b));
        let norm_persistence = if max_pers > 0.0 { 
            self.persistence.mapv(|x| x / max_pers) 
        } else { 
            Array1::zeros(n) 
        };

        // 4. Change (Onset detection via presence - smoothed presence)
        let mut change = Array1::zeros(n);
        for i in 0..n {
            self.smooth_presence[i] = self.change_alpha * self.smooth_presence[i] + (1.0 - self.change_alpha) * presence[i];
            change[i] = (presence[i] - self.smooth_presence[i]).max(0.0);
        }
        let max_change = change.fold(0.0_f64, |a, &b| a.max(b));
        if max_change > 0.0 { 
            change.mapv_inplace(|x| x / max_change); 
        }

        // 5. Continuity (Temporal correlation + frequency smoothness)
        let mut time_coherent = Array1::zeros(n);
        for i in 0..n {
            time_coherent[i] = (presence[i] * self.prev_presence[i]).sqrt();
        }

        let freq_smoothed = self.smooth_along_erb(&presence, self.continuity_erb_sigma);
        let mut freq_coherent = Array1::zeros(n);
        for i in 0..n {
            freq_coherent[i] = (1.0 - (presence[i] - freq_smoothed[i]).abs()).max(0.0);
        }

        let mut continuity = time_coherent * freq_coherent;
        let max_cont = continuity.fold(0.0_f64, |a, &b| a.max(b));
        if max_cont > 0.0 { 
            continuity.mapv_inplace(|x| x / max_cont); 
        }

        // 6. Harmonic Coherence (Stub)
        let coherence: Array1<f64> = Array1::ones(n);

        // Integration (Geometric mean of features integrated with availability)
        let mut field = Array1::zeros(n);
        for i in 0..n {
            let features: f64 = presence[i] * norm_persistence[i] * continuity[i] * change[i] * coherence[i];
            field[i] = availability[i] * features.powf(0.2);
        }

        // Update temporal state for next frame
        self.prev_presence.assign(&presence);

        field
    }

    fn smooth_along_erb(&self, x: &Array1<f64>, sigma_erb: f64) -> Array1<f64> {
        let n = x.len();
        let mut y = Array1::zeros(n);
        let bin_hz = if n > 1 { self.freqs[1] - self.freqs[0] } else { 1.0 };

        for i in 0..n {
            let f = self.freqs[i];
            let erb = 24.7 * (1.0 + 4.37 * f / 1000.0);
            let sigma_hz = sigma_erb * erb;
            let sigma_bins = (sigma_hz / bin_hz).max(1.0);
            
            let half_win = (3.0 * sigma_bins).ceil() as isize;
            let mut sum_w = 0.0;
            let mut val = 0.0;
            
            let start = (i as isize - half_win).max(0) as usize;
            let end = (i as isize + half_win).min(n as isize - 1) as usize;

            for k in start..=end {
                let weight = ( -0.5 * ((k as f64 - i as f64) / sigma_bins).powi(2) ).exp();
                val += x[k] * weight;
                sum_w += weight;
            }
            y[i] = if sum_w > 0.0 { val / sum_w } else { x[i] };
        }
        y
    }
}

File: go/pkg/tracking/affordance.go
package tracking

import (
	"math"
)

// AffordanceField calculates the spectral affordance field for a cappella listening.
type AffordanceField struct {
	Frequencies []float64

	// State for persistence and change
	prevPresence   []float64
	persistence    []float64
	smoothPresence []float64

	// Hyperparameters
	PersistenceAlpha   float64
	ChangeAlpha        float64
	MaskingFloorDB     float64
	MaskingSpreadERB   float64
	DominanceLowHz     float64
	DominanceHighHz    float64
	DominanceWeight    float64
	ContinuityERBSigma float64
}

// NewAffordanceField creates a new AffordanceField with default parameters.
func NewAffordanceField(sampleRate float64, frameSize int) *AffordanceField {
	nBins := frameSize/2 + 1
	freqs := make([]float64, nBins)
	for i := range freqs {
		freqs[i] = float64(i) * sampleRate / float64(frameSize)
	}

	return &AffordanceField{
		Frequencies:        freqs,
		prevPresence:       make([]float64, nBins),
		persistence:        make([]float64, nBins),
		smoothPresence:     make([]float64, nBins),
		PersistenceAlpha:   0.95,
		ChangeAlpha:        0.8,
		MaskingFloorDB:     -60.0,
		MaskingSpreadERB:   1.5,
		DominanceLowHz:     500.0,
		DominanceHighHz:    2000.0,
		DominanceWeight:    1.0,
		ContinuityERBSigma: 0.5,
	}
}

// Update the affordance field with a new magnitude spectrum frame.
func (af *AffordanceField) Update(mag []float64) []float64 {
	n := len(mag)
	if n != len(af.Frequencies) {
		// Handle mismatch if necessary, though they should be consistent
		return nil
	}

	// 1. Feature Presence (normalized magnitude)
	maxMag := 0.0
	for _, v := range mag {
		if v > maxMag {
			maxMag = v
		}
	}
	presence := make([]float64, n)
	if maxMag > 0 {
		for i, v := range mag {
			presence[i] = v / maxMag
		}
	}

	// 2. Peripheral Availability (Masking + Dominance)
	availability := make([]float64, n)
	peakDB := math.Inf(-1)
	magDB := make([]float64, n)
	for i, v := range mag {
		db := 20.0 * math.Log10(v+1e-12)
		magDB[i] = db
		if db > peakDB {
			peakDB = db
		}
	}

	for i := range availability {
		rel := magDB[i] - peakDB
		val := (rel - af.MaskingFloorDB) / (0.0 - af.MaskingFloorDB)
		if val < 0 {
			val = 0
		} else if val > 1 {
			val = 1
		}
		availability[i] = val
	}

	// Apply ERB smoothing to availability
	availability = af.smoothAlongERB(availability, af.MaskingSpreadERB)

	// Dominance weight
	maxAvail := 0.0
	for i := range availability {
		if af.Frequencies[i] >= af.DominanceLowHz && af.Frequencies[i] <= af.DominanceHighHz {
			availability[i] *= 1.0 + af.DominanceWeight
		}
		if availability[i] > maxAvail {
			maxAvail = availability[i]
		}
	}

	// Normalize availability
	if maxAvail > 0 {
		for i := range availability {
			availability[i] /= maxAvail
		}
	}

	// 3. Persistence (Exponential Moving Average)
	maxPers := 0.0
	normPersistence := make([]float64, n)
	for i := range af.persistence {
		af.persistence[i] = af.PersistenceAlpha*af.persistence[i] + (1.0-af.PersistenceAlpha)*presence[i]
		if af.persistence[i] > maxPers {
			maxPers = af.persistence[i]
		}
	}
	if maxPers > 0 {
		for i := range normPersistence {
			normPersistence[i] = af.persistence[i] / maxPers
		}
	}

	// 4. Change (Onset detection via presence - smoothed presence)
	change := make([]float64, n)
	maxChange := 0.0
	for i := range af.smoothPresence {
		af.smoothPresence[i] = af.ChangeAlpha*af.smoothPresence[i] + (1.0-af.ChangeAlpha)*presence[i]
		val := presence[i] - af.smoothPresence[i]
		if val < 0 {
			val = 0
		}
		change[i] = val
		if change[i] > maxChange {
			maxChange = change[i]
		}
	}
	if maxChange > 0 {
		for i := range change {
			change[i] /= maxChange
		}
	}

	// 5. Continuity (Temporal correlation + frequency smoothness)
	timeCoherent := make([]float64, n)
	hasPrev := false
	for _, v := range af.prevPresence {
		if v > 0 {
			hasPrev = true
			break
		}
	}

	for i := range timeCoherent {
		if !hasPrev {
			timeCoherent[i] = presence[i] // Initial frame fallback
		} else {
			timeCoherent[i] = math.Sqrt(presence[i] * af.prevPresence[i])
		}
	}

	freqSmoothed := af.smoothAlongERB(presence, af.ContinuityERBSigma)
	freqCoherent := make([]float64, n)
	for i := range freqCoherent {
		val := 1.0 - math.Abs(presence[i]-freqSmoothed[i])
		if val < 0 {
			val = 0
		}
		freqCoherent[i] = val
	}

	continuity := make([]float64, n)
	maxCont := 0.0
	for i := range continuity {
		continuity[i] = timeCoherent[i] * freqCoherent[i]
		if continuity[i] > maxCont {
			maxCont = continuity[i]
		}
	}
	if maxCont > 0 {
		for i := range continuity {
			continuity[i] /= maxCont
		}
	}

	// 6. Harmonic Coherence (Stub)
	coherence := 1.0

	// Integration (Weighted combination instead of strict geometric mean if features are missing)
	field := make([]float64, n)
	for i := range field {
		// Use a slightly more robust integration for initial frames
		featSum := presence[i] + normPersistence[i] + continuity[i] + change[i]
		if featSum == 0 {
			field[i] = 0
			continue
		}

		// Geometric mean but with small epsilon or fallback
		eps := 0.01
		features := (presence[i] + eps) * (normPersistence[i] + eps) * (continuity[i] + eps) * (change[i] + eps) * coherence
		field[i] = availability[i] * math.Pow(features, 0.2)
	}

	// Update temporal state
	copy(af.prevPresence, presence)

	return field
}

func (af *AffordanceField) smoothAlongERB(x []float64, sigmaERB float64) []float64 {
	n := len(x)
	y := make([]float64, n)
	binHz := 1.0
	if n > 1 {
		binHz = af.Frequencies[1] - af.Frequencies[0]
	}

	for i := 0; i < n; i++ {
		f := af.Frequencies[i]
		erb := 24.7 * (1.0 + 4.37*f/1000.0)
		sigmaHz := sigmaERB * erb
		sigmaBins := sigmaHz / binHz
		if sigmaBins < 1.0 {
			sigmaBins = 1.0
		}

		halfWin := int(math.Ceil(3.0 * sigmaBins))
		sumW := 0.0
		val := 0.0

		start := i - halfWin
		if start < 0 {
			start = 0
		}
		end := i + halfWin
		if end >= n {
			end = n - 1
		}

		for k := start; k <= end; k++ {
			diff := float64(k - i)
			weight := math.Exp(-0.5 * math.Pow(diff/sigmaBins, 2))
			val += x[k] * weight
			sumW += weight
		}
		if sumW > 0 {
			y[i] = val / sumW
		} else {
			y[i] = x[i]
		}
	}
	return y
}
